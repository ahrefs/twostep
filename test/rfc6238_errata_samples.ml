module I = Twostep.Internals

let counters = [
  59L;
  1111111109L;
  1111111111L;
  1234567890L;
  2000000000L;
  20000000000L
]

let algorithms = [ "SHA-1"; "SHA-256"; "SHA-512" ]

let sha1_hex_secret = "3132333435363738393031323334353637383930"
let sha1_secret = Cstruct.to_string @@ Cstruct.of_hex sha1_hex_secret

let sha256_hex_secret = "3132333435363738393031323334353637383930" ^
         "313233343536373839303132"
let sha256_secret = Cstruct.to_string @@ Cstruct.of_hex sha256_hex_secret

let sha512_hex_secret = "3132333435363738393031323334353637383930" ^
         "3132333435363738393031323334353637383930" ^
         "3132333435363738393031323334353637383930" ^
         "31323334"
let sha512_secret = Cstruct.to_string @@ Cstruct.of_hex sha512_hex_secret

let counter_padding =
  counters
  |> Base.List.reduce_exn ~f:Base.Int64.max
  |> Base.Int64.to_string
  |> Base.String.length

let get_secret = function
| "SHA-1"   -> sha1_secret
| "SHA-256" -> sha256_secret
| "SHA-512" -> sha512_secret
| hash      -> failwith ("Unknown hash algorithm: " ^ hash)

let _ =
  Base.List.iter counters ~f:(function counter -> begin
    let text_counter = Base.Int64.to_string counter in
    let padded_counter =
      I.pad ~basis:counter_padding ~byte:' ' ~direction:I.padOnRight text_counter
    in
    let payload_counter = I.counter ~timestamp:(function () -> counter) () in
    let hex_counter =
      Base.String.uppercase @@ Hex.show @@ Hex.of_string payload_counter
    in
    Base.List.iter algorithms ~f:(function algorithm -> begin
      let secret = get_secret algorithm in
      let image = I.hmac ~hash:algorithm ~secret payload_counter in
      let code = I.truncate ~image ~digits:8 in
      print_endline (
        padded_counter ^ " | " ^ hex_counter ^ " | " ^ code ^ " | " ^ algorithm
      )
    end)
  end)
