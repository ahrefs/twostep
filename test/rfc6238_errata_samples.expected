59          | 0000000000000001 | 94287082 | SHA-1
59          | 0000000000000001 | 46119246 | SHA-256
59          | 0000000000000001 | 90693936 | SHA-512
1111111109  | 00000000023523EC | 07081804 | SHA-1
1111111109  | 00000000023523EC | 68084774 | SHA-256
1111111109  | 00000000023523EC | 25091201 | SHA-512
1111111111  | 00000000023523ED | 14050471 | SHA-1
1111111111  | 00000000023523ED | 67062674 | SHA-256
1111111111  | 00000000023523ED | 99943326 | SHA-512
1234567890  | 000000000273EF07 | 89005924 | SHA-1
1234567890  | 000000000273EF07 | 91819424 | SHA-256
1234567890  | 000000000273EF07 | 93441116 | SHA-512
2000000000  | 0000000003F940AA | 69279037 | SHA-1
2000000000  | 0000000003F940AA | 90698825 | SHA-256
2000000000  | 0000000003F940AA | 38618901 | SHA-512
20000000000 | 0000000027BC86AA | 65353130 | SHA-1
20000000000 | 0000000027BC86AA | 77737706 | SHA-256
20000000000 | 0000000027BC86AA | 47863826 | SHA-512
